cpp/arcticdb/version/symbol_list.cpp

```diff
@@ -6,292 +6,703 @@
     }
 
-    SymbolList::CollectionType SymbolList::load_from_symbol_list_keys(
-            const std::shared_ptr<StreamSource>& store,
-            const folly::Range<KeyVectorItr>& keys) {
-        SYMBOL_LIST_RUNTIME_LOG("Loading symbols from symbol list keys");
-        bool read_compaction = false;
-        CollectionType symbols{};
-        for(const auto& key : keys) {
-            if(key.id() == compaction_id) {
-                read_list_from_storage(store, key, symbols);
-                read_compaction = true;
+    if(!problematic_symbols.empty()) {
+        std::map<StreamId, VersionId> symbol_versions;
+        for(const auto& [symbol, reference_pair] : problematic_symbols)
+            symbol_versions.try_emplace(symbol, reference_pair.first);
+
+        auto versions = batch_get_specific_version(store, version_map, symbol_versions, false);
+
+        for (const auto& [symbol, reference_pair] : problematic_symbols) {
+            auto reference_id = reference_pair.first;
+
+            if (auto version = versions->find(symbol); version != versions->end()) {
+                ARCTICDB_DEBUG(log::symbol(), "Problematic symbol/version pair: {}@{}: exists", symbol, reference_id);
+                symbols.emplace_back(symbol, reference_id, version->second.creation_ts(), ActionType::ADD);
             }
             else {
-                const auto& action = key.id();
-                const auto& symbol = key.start_index();
-                if(action == StreamId{DeleteSymbol}) {
-                    ARCTICDB_DEBUG(log::version(), "Got delete action for symbol '{}'", symbol);
-                    symbols.erase(symbol);
-                }
-                else {
-                    ARCTICDB_DEBUG(log::version(), "Got insert action for symbol '{}'", symbol);
-                    symbols.insert(symbol);
-                }
+                ARCTICDB_DEBUG(log::symbol(), "Problematic symbol/version pair: {}@{}: does not exist", symbol, reference_id);
+                symbols.emplace_back(symbol, reference_id, reference_pair.second, ActionType::DELETE);
             }
         }
-        SYMBOL_LIST_RUNTIME_LOG_VAL("Post load, got following number of symbols: {}", symbols.size());
-        if(!read_compaction) {
-            SYMBOL_LIST_RUNTIME_LOG_VAL("Read no compacted segment from symbol list of size: {}", keys.size());
+        std::sort(std::begin(symbols), std::end(symbols), [] (const auto& l, const auto& r) {
+            return l.stream_id_ < r.stream_id_;
+        });
+    }
+
+    return symbols;
+}
+
+CollectionType load_from_symbol_list_keys(
+        const std::shared_ptr<VersionMap>& version_map,
+        const std::shared_ptr<Store>& store,
+        const std::vector<AtomKey>& keys,
+        const MaybeCompaction& maybe_compaction) {
+    ARCTICDB_RUNTIME_DEBUG(log::symbol(),"Loading symbols from symbol list keys");
+    util::check(maybe_compaction.has_value(), "Expected prior compaction when loading from symbol list");
```

Comment:
Not sure. It's a very lightweight check, it documents the expectation of the function and it's robust to any future refactoring, since if someone were to call this from elsewhere and not check they'd get a std::bad_optional_access. Kind of on the fence here, will remove if it really bothers you

Reply:
It won't keep me up at night, I'd just tend to use the type system for this kind of check

Reply:
Sorry, yes I had misunderstood what you were suggesting. I think that's a good idea